expr/fieldexpr.go

```diff
@@ -1,176 +1,101 @@
-		}
+	idx, ok := recType.ColumnOfField(field)
+	if !ok {
+		return zng.Value{}, ErrNoSuchField
 	}
-	// record doesn't have the named field
-	return zng.Value{}
-}
-
-// CompileFieldExpr() takes a FieldExpr AST (which represents either a
-// simple field reference like "fieldname" or something more complex
-// like "fieldname[0].subfield.subsubfield[3]") and compiles it into a
-// FieldExprResolver -- a function that takes a zbuf.Record and extracts the
-// value to which the FieldExpr refers.  If the FieldExpr cannot be
-// compiled, this function returns an error.  If the resolver is given
-// a record for which the given expression cannot be evaluated (e.g.,
-// if the record doesn't have a requested field or an array index is
-// out of bounds), the resolver returns (nil, nil).
-func CompileFieldExpr(node ast.FieldExpr) (FieldExprResolver, error) {
-	var ops []fieldop = make([]fieldop, 0)
-	var field string
-
-	// First collapse the AST to a simple array of fieldop
-	// objects so the actual resolver can run reasonably efficiently.
-outer:
-	for {
-		switch op := node.(type) {
-		case *ast.FieldRead:
-			field = op.Field
-			break outer
-		case *ast.FieldCall:
-			switch op.Fn {
-			// Len handled separately
-			case "Index":
-				idx, err := strconv.ParseInt(op.Param, 10, 64)
-				if err != nil {
-					return nil, err
-				}
-				ops = append([]fieldop{&arrayIndex{idx}}, ops...)
-				node = op.Field
-			case "RecordFieldRead":
-				ops = append([]fieldop{&fieldRead{op.Param}}, ops...)
-				node = op.Field
-			default:
-				return nil, fmt.Errorf("unknown FieldCall: %s", op.Fn)
-			}
-		default:
-			return nil, errors.New("filter AST unknown field op")
-		}
+	typ := recType.Columns[idx].Type
+	if record.Bytes == nil {
+		// Value was unset.  Return unset value of the indicated type.
+		return zng.Value{typ, nil}, nil
 	}
-
-	// Here's the actual resolver: grab the top-level field and then
-	// apply any additional operations.
-	return func(r *zng.Record) zng.Value {
-		col, ok := r.Type.LUT[field]
-		if !ok {
-			// original field doesn't exist
-			return zng.Value{}
-		}
-		e := r.Value(col)
-		for _, op := range ops {
-			e = op.apply(e)
-			if e.Type == nil {
-				break
-			}
-		}
-		return e
-	}, nil
+	zv, err := getNthFromContainer(record.Bytes, uint(idx))
+	if err != nil {
+		return zng.Value{}, err
+	}
+	return zng.Value{recType.Columns[idx].Type, zv}, nil
 }
 
-// CompileFieldExprs calls CompileFieldExpr for each element of nodes.
-func CompileFieldExprs(nodes []ast.FieldExpr) ([]FieldExprResolver, error) {
-	var resolvers []FieldExprResolver
-	if nodes != nil {
-		resolvers = make([]FieldExprResolver, 0, len(nodes))
-		for _, exp := range nodes {
-			res, err := CompileFieldExpr(exp)
-			if err != nil {
-				return nil, err
-			}
-			resolvers = append(resolvers, res)
+func (f *FieldExpr) Eval(rec *zng.Record) (zng.Value, error) {
+	// XXX make test for 'put x=a[selector]' where selector is another field
```

Comment:
Thanks for flagging this.  This is relevant to the index operator not the dot operator and doesn't work yet.  I was planning to add support for this in the zng-beta PR since this is a change to the spec allowing any string to reference a field, e.g., `rec["@f_.oo"]`.  For this PR, I will move this comment to the index operator code path and it will show up in my zng-beta rebase.